source/common/upstream/load_balancer_impl.cc

```diff
@@ -553,18 +566,21 @@ bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
     return true;
   }
 
-  // lb_local_cluster_not_ok is bumped for "Local host set is not set or it is
-  // panic mode for local cluster".
-  if (!host_set.healthyHostsPerLocality().hasLocalLocality() ||
-      host_set.healthyHostsPerLocality().get()[0].empty()) {
-    stats_.lb_local_cluster_not_ok_.inc();
+  // Do not perform locality routing if there are too few local localities for zone routing to have
+  // an effect.
+  if (localHostSet().hostsPerLocality().get().size() < 2) {
     return true;
   }
 
-  // Same number of localities should be for local and upstream cluster.
-  if (host_set.healthyHostsPerLocality().get().size() !=
-      localHostSet().healthyHostsPerLocality().get().size()) {
-    stats_.lb_zone_number_differs_.inc();
+  // Do not perform locality routing if the local cluster doesn't have any hosts in the current
+  // envoy's local locality. This breaks our assumptions about the local cluster being correctly
+  // configured, so we don't have enough information to perform locality routing. Note: If other
+  // envoys do exist according to the local cluster, they will still be able to perform locality
+  // routing correctly. This will not cause a traffic imbalance because other envoys will not know
+  // about the current one, so they will not factor it into locality routing calculations.
+  if (!localHostSet().hostsPerLocality().hasLocalLocality() ||
+      localHostSet().hostsPerLocality().get()[0].empty()) {
+    stats_.lb_local_cluster_not_ok_.inc();
     return true;
   }
```

Comment:
`earlyExitNonLocalityRouting` needs a few minor changes:
- No longer need to disable zone-aware routing if the zone number differs, because this is now supported
- We now support the case where the upstream host set doesn't have any hosts in the local zone

I've also added an "easy out" for when zone-aware routing has no effect:
- if there are 1 or fewer local zones, then zone-aware routing won't ever change our routing decisions, so don't bother